pkg/subscribe/watcher_test.go

```diff
@@ -0,0 +1,164 @@
+package subscribe
+
+import (
+	"context"
+	"net/http"
+	"testing"
+
+	"github.com/rancher/apiserver/pkg/types"
+	"github.com/rancher/wrangler/pkg/schemas"
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_stream(t *testing.T) {
+	tests := []struct {
+		name           string
+		sub            Subscribe
+		wantStartEvent types.APIEvent
+	}{
+		{
+			name: "stream all",
+			sub: Subscribe{
+				ResourceType: "test",
+			},
+			wantStartEvent: types.APIEvent{
+				Name:         "resource.start",
+				ResourceType: "test",
+			},
+		},
+		{
+			name: "stream by namespace",
+			sub: Subscribe{
+				ResourceType: "test",
+				Namespace:    "test-ns",
+			},
+			wantStartEvent: types.APIEvent{
+				Name:         "resource.start",
+				ResourceType: "test",
+				Namespace:    "test-ns",
+			},
+		},
+		{
+			name: "stream by selector",
+			sub: Subscribe{
+				ResourceType: "test",
+				Selector:     "foo=bar",
+			},
+			wantStartEvent: types.APIEvent{
+				Name:         "resource.start",
+				ResourceType: "test",
+				Selector:     "foo=bar",
+			},
+		},
+		{
+			name: "stream by id",
+			sub: Subscribe{
+				ResourceType: "test",
+				ID:           "test-resource",
+			},
+			wantStartEvent: types.APIEvent{
+				Name:         "resource.start",
+				ResourceType: "test",
+				ID:           "test-resource",
+			},
+		},
+	}
+	ws := WatchSession{
+		apiOp: &types.APIRequest{
+			Name: "test",
+			Schemas: &types.APISchemas{
+				Schemas: map[string]*types.APISchema{
+					"test": {
+						Schema: &schemas.Schema{
+							ID: "test",
+						},
+						Store: &mockStore{},
+					},
+				},
+			},
+			AccessControl: &mockAC{},
+			Request:       &http.Request{},
+		},
+		getter: DefaultGetter,
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			result := make(chan types.APIEvent, 1)
+			err := ws.stream(context.TODO(), test.sub, result)
+			assert.Nil(t, err)
+			done := make(chan bool)
+			var gotEvent types.APIEvent
+			go func() {
+				gotEvent = <-result
+				done <- true
+			}()
+			<-done
```

Comment:
Added a select here for <- done. I don't think I can block on <-result, that needs to be in its own goroutine and we need something else to signal that the goroutine is finished.

Reply:
As long as the test doesn't stall (ran on my end with `go test -race` and looked fine), no issue on my end. Will leave it up to @KevinJoiner to resolve since he opened the thread first (though he's already approved so I think he's ok with it).